contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+	if( interpolationToken == pxr::UsdGeomTokens->uniform )
+	{
+		return IECore::PrimitiveVariable::Uniform;
+	}
+	if( interpolationToken == pxr::UsdGeomTokens->faceVarying )
+	{
+		return IECore::PrimitiveVariable::FaceVarying;
+	}
+	if( interpolationToken == pxr::UsdGeomTokens->constant )
+	{
+		return IECore::PrimitiveVariable::Constant;
+	}
+
+	return IECore::PrimitiveVariable::Invalid;
+}
+
+template<typename DestElementType, typename SourceElementType>
+typename IECore::GeometricTypedData<std::vector<DestElementType> >::Ptr convertIt( const pxr::VtValue &value )
+{
+	typedef pxr::VtArray<SourceElementType> SourceArrayType;
+	typedef IECore::GeometricTypedData<std::vector<DestElementType> > DestArrayType;
+	typedef std::vector<DestElementType> DestStorageType;
+
+	if( value.IsHolding<SourceArrayType>() )
+	{
+		const auto &r = value.Get<SourceArrayType>();
+
+		typename DestArrayType::Ptr d = new DestArrayType();
+		DestStorageType &t = d->writable();
+
+		t.resize( r.size() );
+
+		for( size_t i = 0; i < r.size(); ++i )
+		{
+			t[i] = convert( r[i] );
+		}
+
+		return d;
+	}
+
+	return nullptr;
+}
+
+IECore::PrimitiveVariable fixForNow( const IECore::PrimitiveVariable &srcPrimVar )
+{
+	if( srcPrimVar.indices )
+	{
+		return IECore::PrimitiveVariable( srcPrimVar.interpolation, srcPrimVar.expandedData() );
+	}
+	else
+	{
+		return srcPrimVar;
+	}
+}
+
+std::string cleanPrimVarName( const std::string &primVarName )
+{
+	return boost::algorithm::replace_first_copy( primVarName, "primvars:", "" );
+}
+
+//!
+struct PrimVarConverter
+{
+	PrimVarConverter(IECore::PrimitivePtr primitive, const pxr::UsdGeomPrimvar& primVar, pxr::UsdTimeCode time): m_primitive(primitive), m_primVar(primVar), m_time(time), m_converted(false) {}
+
+	template< typename U > void operator()(U x)
+	{
+		if ( m_converted )
+			return;
+
+		pxr::VtValue value;
+		m_primVar.Get( &value, m_time );
+
+		auto p = convertIt<typename U::second, typename U::first>( value );
+
+		if ( !p )
+			return;
+
+		IECore::PrimitiveVariable::Interpolation interpolation = convertInterpolation( m_primVar.GetInterpolation() );
+
+		pxr::VtIntArray srcIndices;
+		m_primVar.GetIndices( &srcIndices, m_time );
+
+		auto indices = !srcIndices.empty() ? convert( srcIndices ) : nullptr;
+
+		std::string cleanedPrimvarName = cleanPrimVarName( m_primVar.GetName() );
+		m_primitive->variables[cleanedPrimvarName] = fixForNow( IECore::PrimitiveVariable( interpolation, p, indices ) );
+		m_converted = true;
+	}
+
+	IECore::PrimitivePtr m_primitive;
+	const pxr::UsdGeomPrimvar& m_primVar;
+	pxr::UsdTimeCode m_time;
+	bool m_converted;
+};
+
+void convertPrimVar( IECore::PrimitivePtr primitive, pxr::UsdTimeCode time, const pxr::UsdGeomPrimvar &primVar )
+{
+	IECore::PrimitiveVariable::Interpolation interpolation = convertInterpolation( primVar.GetInterpolation() );
+	if( interpolation == IECore::PrimitiveVariable::Invalid )
+	{
+		return;
+	}
+
+	PrimVarConverter primVarConverter( primitive, primVar, time );
```

Comment:
Not sure if this is nifty or terrible but I've used something like this in the VDB code to dispatch to concrete grid types. Let me know what you think and I'll adjust accordingly.